tezos/wrapper/src/service.rs

```diff
@@ -462,25 +441,99 @@ impl Drop for ProtocolController<'_> {
     }
 }
 
+/// Endpoint consists of a protocol runner and IPC communication (command and event channels).
+pub struct ProtocolRunnerEndpoint<Runner: ProtocolRunner> {
+    pub name: String,
+    runner: Runner,
+    log: Logger,
+
+    pub commands: IpcCmdServer,
+    pub events: IpcEvtServer,
+}
+
+impl<Runner: ProtocolRunner + 'static> ProtocolRunnerEndpoint<Runner> {
+    pub fn new(name: &str, configuration: ProtocolEndpointConfiguration, log: Logger) -> ProtocolRunnerEndpoint<Runner> {
+        let evt_server = IpcEvtServer::new();
+        let cmd_server = IpcCmdServer::new(configuration.clone());
+        ProtocolRunnerEndpoint {
+            name: name.to_string(),
+            runner: Runner::new(configuration, cmd_server.0.client().path(), evt_server.0.client().path()),
+            commands: cmd_server,
+            events: evt_server,
+            log,
+        }
+    }
+
+    /// Starts protocol runner sub-process just once and you can take care of it
+    pub fn start(&self) -> Result<Child, ProtocolServiceError> {
+        info!(self.log, "Starting protocol runner process"; "endpoint" => self.name.clone());
+        self.runner.spawn()
+    }
+
+    /// Starts protocol runner sub-process and takes care of it automatically.
+    /// If sub-process failed, it is automatically spawned another sub-process.
+    /// Returns AtomicBool, if set to false, than terminates sub-process
+    pub fn start_in_restarting_mode(&mut self) -> Result<Arc<AtomicBool>, ProtocolServiceError> {
```

Comment:
I think we could utilize Riker actor system here and save one thread

Reply:
I think so also, but it was not so easy on the first look :), and also in our case, ProtocolEndpoint is splitted beetween two actors (ContextListener, ChainFeeder), so I just refactored watching thread to be able easily reused the same code for another ProtocolEndpoint at first step